### PR TITLE
HPCC-20368 Certificates defaults set correctly for developers

### PR DESCRIPTION
**Pull request by Michael-Gardner**
… to use them.

Signed-off-by: Michael Gardner <michael.gardner@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@mckellyln Please take a look at this when you get a chance.

File: initfiles/etc/DIR_NAME/environment.xml.in [hunk 1]
```diff
@@ -274,16 +274,16 @@
...
+          certificateFileName="@HOME_DIR@/@RUNTIME_USER@/certificate/certificate.pem"
           city=""
           country="US"
           daysValid="365"
           enableVerification="false"
           organization="Customer of HPCCSystems"
           organizationalUnit=""
-          passphrase=""
-          privateKeyFileName="privatekey.cer"
+          passphrase="hpccsystems-platform"
```
Comment (mckellyln) [inline initfiles/etc/DIR_NAME/environment.xml.in]: Isn't passphrase saved as md5 encrypted ?

File: initfiles/etc/DIR_NAME/environment.xml.in [hunk 2]
```diff
@@ -274,16 +274,16 @@
              name="s1"
              netAddress="."/>
    <HTTPS acceptSelfSigned="true"
-          CA_Certificates_Path="ca.pem"
-          certificateFileName="certificate.cer"
+          CA_Certificates_Path="@HOME_DIR@/@RUNTIME_USER@/certificate"
```
[the remark this answers is not in the document]
Comment (Michael-Gardner) [reply initfiles/etc/DIR_NAME/environment.xml.in]: Checked with Russ, and I think the entire file path in the 'fileName' attributes was wrong.  corrected it.

File: initfiles/componentfiles/configxml/esp.xsd.in [hunk 3]
```diff
@@ -475,21 +475,21 @@
...
                         </xs:attribute>
-                        <xs:attribute name="privateKeyFileName" type="xs:string" use="optional" default="@HOME_DIR@/@RUNTIME_USER@/certificate/key.pem">
+                        <xs:attribute name="privateKeyFileName" type="xs:string" use="optional" default="key.pem">
                             <xs:annotation>
                                 <xs:appinfo>
                                     <tooltip>Name of destination file in which the private key will be written.</tooltip>
                                 </xs:appinfo>
                             </xs:annotation>
                         </xs:attribute>
-                        <xs:attribute name="passphrase" type="xs:string" use="optional" default="hpccsystems-platform">
```
Comment (Michael-Gardner) [inline initfiles/componentfiles/configxml/esp.xsd.in]: Not certain if I should keep this as a default here.  Thoughts ?

**Comment by mckellyln** [comment]
I am getting this error:
```
00000053 2018-09-05 16:34:47.484  3007  3007 "ERROR: ESP Unhandled IException (-1 -- error loading certificate file certificate.pem - error  :02001002:system library:fopen:No such file or directory)"
```
with this default xml file
```
   <HTTPS acceptSelfSigned="true"
          CA_Certificates_Path="/home/hpcc/certificate"
          certificateFileName="certificate.pem"
          city=""
          country="US"
          daysValid="365"
          enableVerification="false"
          organization="Customer of HPCCSystems"
          organizationalUnit=""
          passphrase="XC+0Byt45EYD6qV5++8HW+lgsYB/nGud8FtgrubKdLk="
          privateKeyFileName="key.pem"
          regenerateCredentials="false"
          requireAddressMatch="false"
          state=""
          trustedPeers="anyone"/>
```

**Comment by mckellyln** [comment]
But esp starts ok with this xml file
```
   <HTTPS acceptSelfSigned="true"
          CA_Certificates_Path="/does/not/matter"
          certificateFileName="/home/hpcc/certificate/certificate.pem"
          city=""
          country="US"
          daysValid="365"
          enableVerification="false"
          organization="Customer of HPCCSystems"
          organizationalUnit=""
          passphrase="XC+0Byt45EYD6qV5++8HW+lgsYB/nGud8FtgrubKdLk="
          privateKeyFileName="/home/hpcc/certificate/key.pem"
          regenerateCredentials="false"
          requireAddressMatch="false"
          state=""
          trustedPeers="anyone"/>
```

**Comment by mckellyln** [comment]
So it would seem CA_Certificates_Path is not used to locate the certificateFileName and perhaps also the privateKeyFileName

**Comment by mckellyln** [comment]
@Michael-Gardner I think go back to use full path for privateKeyFileName and certificateFileName and I suppose put CA_Certificates_Path back to what it was (or remove it if it really is not used anywhere)

**Comment by mckellyln** [comment]
I just tried it and am getting an error -
```
00000055 2018-09-07 09:13:04.121 21281 21281 "ERROR: ESP Unhandled IException (-1 -- error loading private key file /home/hpcc/certificate/key.pem - error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt)"
```

**Comment by mckellyln** [comment]
xml is:
```
   <HTTPS acceptSelfSigned="true"
          CA_Certificates_Path="/home/hpcc/certificate"
          certificateFileName="/home/hpcc/certificate/certificate.pem"
          city=""
          country="US"
          daysValid="365"
          enableVerification="false"
          organization="Customer of HPCCSystems"
          organizationalUnit=""
          passphrase="XC+0Byt45EYD6qV5++8HW+lgsYB/nGud8FtgrubKdLk="
          privateKeyFileName="/home/hpcc/certificate/key.pem"
          regenerateCredentials="false"
          requireAddressMatch="false"
          state=""
          trustedPeers="anyone"/>
```

**Comment by mckellyln** [comment]
@Michael-Gardner I think this means the key.pem file is encrypted and the code is not handling it.
I am lost as to why it started ok before, unless there was a change to how the key.pem file is generated in the last round of commits merged into master ?

**Comment by mckellyln** [comment]
@Michael-Gardner I think this means the key.pem is encrypted and the passphrase has not been used properly to decrypt it.

**Comment by mckellyln** [comment]
@Michael-Gardner can you test as well to see if you get the same issue ?

**Comment by Michael-Gardner** [comment]
@mckellyln I asked Russ about it since he knows more about the https functionality than I do, and he said it was likely caused by an incorrect file format.  I remember something similar happening in the work we did for defaulting https before we rolled it all back for 7.0, but it's a bit out of the scope of this PR.

File: initfiles/etc/DIR_NAME/environment.xml.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (mckellyln) [reply initfiles/etc/DIR_NAME/environment.xml.in]: @Michael-Gardner I would put this back to what it was "ca.pem"

**Comment by mckellyln** [comment]
@Michael-Gardner I would restore 
```
CA_Certificates_Path="ca.pem"
```
But otherwise it is good.

**Comment by mckellyln** [comment]
@Michael-Gardner @richardkchapman @jakesmith I think the reason it worked a few days ago and not now is that the passphrase decrypt code may have just changed from jlib::aesDecrypt() to cryptohelper::aesDecrypt() ?

**Comment by mckellyln** [comment]
@jakesmith I changed jencrypt.cpp to se jlib versions of encrypt() and decrypt() and it now works ok.

**Comment by mckellyln** [comment]
ok, with https://github.com/hpcc-systems/HPCC-Platform/pull/11678 it works ok.
@Michael-Gardner I would just revert -
```
CA_Certificates_Path="ca.pem"
```
and then it is good to merge

**Comment by Michael-Gardner** [comment]
@mckellyln reverted that one change

**Review by mckellyln — APPROVED** [review]
Looks good to merge

**Comment by mckellyln** [comment]
@richardkchapman ok to merge

**Pull request merged**